nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/CountText.java

```diff
@@ -0,0 +1,318 @@
+                        if (line.trim().length() > 0) {
+                            lineNonEmptyCount++;
+                        }
+                    }
+
+                    if (countWords) {
+                        wordCount += countWordsInLine(line, splitWordsOnSymbols);
+                    }
+
+                    if (countCharacters) {
+                        characterCount += line.length();
+                    }
+                }
+                long stop = System.nanoTime();
+                if (getLogger().isDebugEnabled()) {
+                    final long durationNanos = stop - start;
+                    DecimalFormat df = new DecimalFormat("#.###");
+                    getLogger().debug("Computed metrics in " + durationNanos + " nanoseconds (" + df.format(durationNanos / 1_000_000_000.0) + " seconds).");
+                }
+                String message = generateMetricsMessage();
+                getLogger().info(message);
+            } catch (IllegalStateException e) {
+                error.set(true);
+                getLogger().error(e.getMessage() + " Routing to failure.", e);
+            }
+        });
+
+        if (error.get()) {
+            processSession.transfer(sourceFlowFile, REL_FAILURE);
+        } else {
+            Map<String, String> metricAttributes = new HashMap<>();
+            if (countLines) {
+                metricAttributes.put(TEXT_LINE_COUNT, String.valueOf(lineCount));
+            }
+            if (countLinesNonEmpty) {
+                metricAttributes.put(TEXT_LINE_NONEMPTY_COUNT, String.valueOf(lineNonEmptyCount));
+            }
+            if (countWords) {
+                metricAttributes.put(TEXT_WORD_COUNT, String.valueOf(wordCount));
+            }
+            if (countCharacters) {
+                metricAttributes.put(TEXT_CHARACTER_COUNT, String.valueOf(characterCount));
+            }
+            FlowFile updatedFlowFile = processSession.putAllAttributes(sourceFlowFile, metricAttributes);
+            processSession.transfer(updatedFlowFile, REL_SUCCESS);
+        }
+    }
+
+    private String generateMetricsMessage() {
+        StringBuilder sb = new StringBuilder("Counted ");
+        List<String> metrics = new ArrayList<>();
+        if (countLines) {
+            metrics.add(lineCount + " lines");
+        }
+        if (countLinesNonEmpty) {
+            metrics.add(lineNonEmptyCount + " non-empty lines");
+        }
+        if (countWords) {
+            metrics.add(wordCount + " words");
+        }
+        if (countCharacters) {
+            metrics.add(characterCount + " characters");
+        }
+        sb.append(StringUtils.join(metrics, ", "));
+        return sb.toString();
+    }
+
+    private int countWordsInLine(String line, boolean splitWordsOnSymbols) {
+        if (line == null || line.trim().length() == 0) {
+            return 0;
+        } else {
+            String regex = splitWordsOnSymbols ? SYMBOL_REGEX : WHITESPACE_ONLY_REGEX;
```

Comment:
We should pre-compile these regexes, and a Pattern object as the member variable, then call regex.split(line) -- this way we don't re-compile the regex for every line and instead only compile it once for the processor lifecycle.

Reply:
Good idea.